contracts/StableJoeStaking.sol

```diff
@@ -0,0 +1,312 @@
+
+    /// @notice Emitted when a user emergency withdraws its JOE
+    event EmergencyWithdraw(address indexed user, uint256 amount);
+
+    /**
+     * @notice Initialize a new StableJoeStaking contract
+     * @dev This contract needs to receive an ERC20 `_rewardToken` in order to distribute them
+     * (with JoeMakerV4 in our case)
+     * @param _rewardToken The address of the ERC20 reward token
+     * @param _joe The address of the JOE token
+     * @param _depositFeePercent The deposit fee percent, scalled to 1e18, e.g. 3% is 3e16
+     */
+    function initialize(
+        IERC20Upgradeable _rewardToken,
+        IERC20Upgradeable _joe,
+        address _feeCollector,
+        uint256 _depositFeePercent
+    ) external initializer {
+        __Ownable_init();
+        require(_feeCollector != address(0), "StableJoeStaking: fee collector can't be address 0");
+        require(_depositFeePercent < 5e17, "StableJoeStaking: max deposit fee can't be greater than 50%");
+
+        joe = _joe;
+        depositFeePercent = _depositFeePercent;
+        feeCollector = _feeCollector;
+
+        isRewardToken[_rewardToken] = true;
+        rewardTokens.push(_rewardToken);
+        PRECISION = 1e24;
+    }
+
+    /**
+     * @notice Deposit JOE for reward token allocation
+     * @param _amount The amount of JOE to deposit
+     */
+    function deposit(uint256 _amount) external {
+        UserInfo storage user = userInfo[msg.sender];
+
+        uint256 fee = (_amount * depositFeePercent) / 1e18;
+        uint256 amountMinusFee = _amount.sub(fee);
+
+        uint256 previousAmount = user.amount;
+        uint256 newAmount = user.amount.add(amountMinusFee);
+        user.amount = newAmount;
+
+        uint256 len = rewardTokens.length;
+        uint256 pending;
+        for (uint256 i; i < len; i++) {
+            IERC20Upgradeable token = rewardTokens[i];
+            updateReward(token);
+
+            if (previousAmount > 0) {
+                pending = previousAmount.mul(accRewardPerShare[token]).div(PRECISION).sub(user.rewardDebt[token]);
+                if (pending != 0) safeTokenTransfer(token, msg.sender, pending);
+            }
+            user.rewardDebt[token] = newAmount.mul(accRewardPerShare[token]).div(PRECISION);
+        }
+
+        internalJoeBalance = internalJoeBalance.add(amountMinusFee);
+        joe.safeTransferFrom(msg.sender, feeCollector, fee);
+        joe.safeTransferFrom(msg.sender, address(this), amountMinusFee);
+        emit Deposit(msg.sender, amountMinusFee);
+    }
+
+    /**
+     * @notice get user info
+     * @param _user The address of the user
+     * @param _rewardToken The address of the reward token
+     * @return uint256 the user's amount
+     * @return uint256 the user's reward debt of the reward token chosen
+     */
+    function getUserInfo(address _user, IERC20Upgradeable _rewardToken) external view returns (uint256, uint256) {
+        UserInfo storage user = userInfo[_user];
```

Comment:
view so you can use memory here.

Reply:
I thought too, but actually you cant:
`Member "rewardDebt" is not available in struct StableJoeStaking.UserInfo memory outside of storage.`
I think you can't get a mapping using `memory`

Reply:
TIL

Reply:
Interesting